### ISSUE TITLE
2d bin labels

### ISSUE THREAD

**Issue by schrum2** [schrum2/MM-NEAT#780]
Create MinecraftMAPElitesBlockCountEmptyPercentBinLabels.

One dimension is the block count, which we already have.

Dimensions correspond to fitness functions, so you'll need to make a new one for the second dimension. It measures the percentage of empty space in the shape, but not the entire x y z range. For example, a 2 x 2 x 2 block of quartz has 0 empty space. We only count an AIR block as empty space if it is between the min and max block inclusive it each dimension. 

For binning, divide into 10 percent ranges

**Comment by MuellMark** [comment]
The first step here is creating a new fitness function that counts up the negative space within a shape. It's important to reiterate that this does not mean all air blocks in a shape allotted size, but rather the air present in the shape itself. So, for a 2x2x2 structure generated in a 3x3x3 space, there would still be 0 negative space, despite the 19 air blocks that would surround the shape. From there, we'll see if we want to use the percentages or the integers themselves (or choosing between them) for the other coordinate

**Comment by MuellMark** [comment]
Unit testing for negativeSpaceeCountFitness is complete and passes the tests. Now I will start working on MinecraftMAPElitesBlockCountEmptyPercentBinLabels which will generate the archive in 2D with the block count on one axis, and the negative space count on the other. for a 10x10x10 generation, it should 1001X999 blocks .

**Comment by MuellMark** [comment]
![image](https://user-images.githubusercontent.com/92736704/173136386-d51e65de-b6e6-4809-92fd-b04ac56d91f1.png)
plot generated after verifying unit tests were correct
![2022-06-10_14 16 51](https://user-images.githubusercontent.com/92736704/173136438-8483517c-7f23-4964-bb67-15ca1d75b2c1.png)

**Comment by MuellMark** [comment]
Currently, the fitness function read in the air blocks, but the cppn doesn't generate any air blocks. So when it reads in the block list, no air blocks are present, making all finesses 0. I need to change how it reads in the blocks, so that instead of reading air blocks, it reads in non-air blocks , and subtracts itself from the largest possible shape.

**Comment by schrum2** [comment]
Have a unit test that fills a 6x6x6 area with AIR blocks but has some non-AIR blocks occupying a 4x2x3 region within it, but not completely filling it.

**Comment by MuellMark** [comment]
As of right now, the new unit test was added and they all passed, so I uncommented everything out. This should mean that the negative space count fitness file is working. On top of that, I ran the new characterization with the 2D archive and it looks like it's working, although there are a lot of empty spaces. I believe the next point from here is to make eliminate some, if not all, of the empty archive spaces

**Comment by MuellMark** [comment]
Pictures from a 3x3x3 archive. It was able to achieve max fitness!
![2022-06-13_10 21 48](https://user-images.githubusercontent.com/92736704/173391225-ae454977-f2cd-4658-bc65-0f498f62fdef.png)
![2022-06-13_10 37 46](https://user-images.githubusercontent.com/92736704/173391305-1671d7af-1a1b-4e88-b412-18f07660b924.png)
![2022-06-13_10 30 43](https://user-images.githubusercontent.com/92736704/173391430-e028a3a0-2492-42f5-9ac8-8ddcdd32325e.png)

**Comment by MuellMark** [comment]
At this point, the empty fences that would hole blocks with 0 block count are no longer displayed, and I also fixed the issue with there being one extra row of fences for the negative block type. Now, when the loner shape task runs(with the right parameters), it produces a 2D archive with the block count on one axis, and the negative space count on the other. It is worth noting that some combinations are impossible, such as when 5 blocks are placed, there is no possible way to have 0 negative space, as one block will always jut out at one side. This explains why there are so many empty spaces in the pictures. Now that we have been able to effectively test the 2D archive, #771 should also be able to be closed. 
![2022-06-13_11 51 24](https://user-images.githubusercontent.com/92736704/173404792-76fd7c58-cd45-495d-b593-70d72e0ae46c.png)
